### PR TITLE
feat: add initial create custom view modal

### PR DESCRIPTION
**Pull request by joaoviana**
Closes: #11722

### Description:

- Updates Create custom view modal d
- Allows the option to 'write back to dbt'

> [!NOTE]
> No actual PR is open with this there's some TODOs in the code to address that

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    const form = useForm<FormValues>({
+        initialValues: {
+            name: '',
+            customViewType: CustomViewType.VIRTUAL,
+        },
+        validate: zodResolver(validationSchema),
+    });
+
+    const { data: project } = useProject(projectUuid);
+    const canWriteToDbtProject = !!(
```
Comment (joaoviana) [inline packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:67]: this condition is correct, right @rephus ?

File: packages/common/src/types/explore.ts [hunk 2]
```diff
@@ -90,3 +90,8 @@ export type Table = TableBase & {
     lineageGraph: LineageGraph; // DAG structure representing the lineage of the table
     source?: Source;
 };
+
+export enum CustomViewType {
+    VIRTUAL = 'virtual',
+    PERSISTENT = 'persistent',
```
Comment (joaoviana) [inline packages/common/src/types/explore.ts:96]: what do you think?

**Comment by rephus** [comment]
I don't think this option is available in render, can we enable ? I'll test it locally in the meantime

image

File: packages/common/src/types/explore.ts (hunk 2 above)
Comment (rephus) [reply packages/common/src/types/explore.ts:96]: or perhaps `write_back`

**Comment by rephus** [comment]
Minor UI thing (perhaps for later) , we sh ould remove duplicated / on the path

image

For context, when using github API, the path can't start with /, so it will be `models/file.sql`

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        project?.dbtConnection.type === DbtProjectType.GITHUB
+    );
+    // TODO: Check if possible to create a branch and PR - if not, and it's not a git project, disable option with tooltip
```
Comment (rephus) [inline packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:70]: First you can check `feature flag`, then `    "hasGithub": true,` on health, then you can check if the user has logged in in github with `/api/v1/github/repos/list`

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                        <Stack spacing="xs" pl="xs">
+                            <Text fw={500}>
+                                Files to be created in{' '}
+                                <Badge
+                                    radius="md"
+                                    variant="light"
+                                    color="gray.9"
+                                    fz="xs"
+                                >
+                                    REPO NAME
```
Comment (rephus) [inline packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:197]: We can build a new endpoint to get this and the file path , or you can take this from `api/v1/github/repos/list`

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    );
+    // TODO: Check if possible to create a branch and PR - if not, and it's not a git project, disable option with tooltip
+    const projectDirectory = useMemo(
+        () =>
+            canWriteToDbtProject
+                ? getProjectDirectory(project?.dbtConnection)
+                : undefined,
+        [project?.dbtConnection, canWriteToDbtProject],
+    );
+    const basePathForDbtCustomView = `${projectDirectory}/models/lightdash`;
```
Comment (rephus) [inline packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:78]: Replace multiple / with a single / , perhaps with `.replace(/\/{2,}/g, '/');`?

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx (hunk 4 above)
Comment (rephus) [reply packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:197]: We could make this a Link that opens the Repo in a separate tab

**Review by rephus — APPROVED** [review]
Looks good to me !

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx (hunk 3 above)
Comment (joaoviana) [reply packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:70]: What feature flag do you want me to check here @rephus ?

On the login-check, I've created a separate ticket: https://github.com/lightdash/lightdash/issues/11742

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx (hunk 4 above)
Comment (joaoviana) [reply packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx:197]: Added a TODO

File: packages/common/src/types/explore.ts (hunk 2 above)
Comment (joaoviana) [reply packages/common/src/types/explore.ts]: done! https://github.com/lightdash/lightdash/pull/11728/commits/8d7ff6a055271e6a7c4de0272998a22f976e5f48

File: packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx (hunk 5 above)
Comment (joaoviana) [reply packages/frontend/src/features/sqlRunner/components/SaveCustomViewModal.tsx]: this is what I have now

image


should I remove the first `/`?

**Pull request merged**